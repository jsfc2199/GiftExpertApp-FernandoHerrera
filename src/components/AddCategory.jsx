import PropTypes from "prop-types";
import { useState } from "react";


export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  const onInputChange = (event) => {
    setInputValue(event.target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();

    if (inputValue.trim().length <= 1) return;

    //usamos el callback para tener las categorias actuales
    // setCategories((categories) => [inputValue, ...categories]); 

    //lo anterior valido pero lo haremos solo emitiendo el valor
    onNewCategory(inputValue.trim()) //emitimos el valor

    setInputValue("");
  };

  //onChange: metodo para detectar cambios en inputs
  //agregamos aria-label para las pruebas unitarias del formulario como tal
  return (
    <form onSubmit={(event) => onSubmit(event)} aria-label="form">
      <input
        type="text"
        placeholder="Buscar Gifts"
        value={inputValue}
        onChange={(event) => onInputChange(event)}
      ></input>
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired
}
