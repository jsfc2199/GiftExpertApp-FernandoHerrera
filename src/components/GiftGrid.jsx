//todo usar use effect

import { useEffect, useState } from "react";
import { getGifts } from "../helpers/getGifts";

// eslint-disable-next-line react/prop-types
export const GiftGrid = ({ category }) => {
  const [images, setImages] = useState([]);

  //forma2 para setear las imagenes
  const getImages = async () => {
    const newImages = await getGifts(category);
    setImages(newImages);
  };
  /*
    El useEffect usa 2 argumentos, un callback o una arrow function en este caso de lo que queremos ejecutar
    El segundo argumento es un arreglo de las dependencias que son las condiciones sobre las cuales queremos volver a ejecutar el callback, si este argumento es vaciÃ³ significa que el hook se va a disparar solo la primera vez que se crea y construye el componente

    Si no se tienen dependencias se puede remover el arreglo

    Al insertar una categorua desde AddCategory, el componente se esta construyendo otra vez, entonces es correcto el funcionamiento
    */
  useEffect(() => {
    //forma 1 como es una promesa podemos usar el then
    // getGifts(category)
    // .then(images => setImages(images));
    getImages(category);
  });

  return (
    <>
      <h3>{category}</h3>
      <ol>
        {images.map((id, title) => {
          return <li key={id}>{title} </li>;
        })}
      </ol>
    </>
  );
};
