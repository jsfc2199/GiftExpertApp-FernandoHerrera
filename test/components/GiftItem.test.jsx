/* eslint-disable no-undef */

import { render, screen } from "@testing-library/react";
import { GiftItem } from "./../../src/components/GiftItem";

describe("Pruebas en gift item", () => {
  const title = "Title";
  const url = "https://url/";
  test("should hacer match con el snapshot", () => {
    const { container } = render(<GiftItem title={title} url={url} />);

    expect(container).toMatchSnapshot();
  });

  test("should mostrar la imagen con la url y el ALT indicado", () => {
    render(<GiftItem title={title} url={url} />);

    //forma1
    expect(screen.getByRole("img").src).toBe(url);
    expect(screen.getByRole("img").alt).toBe(title);

    //forma2 igual a lo anterior pero desestructurando
    const { src, alt } = screen.getByRole("img");
    expect(src).toBe(url);
    expect(alt).toBe(title);
  });

  test("should mostrar el titulo como un texto dentro del componente", () => {
    render(<GiftItem title={title} url={url} />);

    expect(screen.getByText(title)).toBeTruthy()
  });
});
