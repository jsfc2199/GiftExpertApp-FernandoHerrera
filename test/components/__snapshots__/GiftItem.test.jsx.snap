// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en gift item should hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Title"
      src="https://url/"
    />
    <p>
      Title
    </p>
  </div>
</div>
`;
